Configuring kafka.message.iterator.className in kafka-2.0.0 profile

The `com.pinterest.secor.reader.LegacyKafkaMessageIterator` is not included when using `mvn -Pkafka-2.0.0`, but it is still mentioned in the default `secor.common.properties`. Due to the way Apache Commons PropertiesConfiguration works it is not possible (?) to _overwrite_ this setting, but rather setting in a configuration that includes `secor.common.properties` directly or indirectly it will just append the value, leading to:
~~~~
Nov 09, 2018 3:06:38 PM com.twitter.ostrich.admin.BackgroundProcess start
INFO: Starting LatchedStatsListener
Nov 09, 2018 3:06:38 PM com.twitter.ostrich.admin.AdminHttpService start
INFO: Admin HTTP interface started on port 9998.
2018-11-09 15:06:38,255 [main] (com.pinterest.secor.common.OstrichAdminService:68) INFO  build.properties build_revision: 7d253c1a5ad1d438990e0362785a7a42346ce625
2018-11-09 15:06:38,267 [main] (com.pinterest.secor.main.ConsumerMain:69) INFO  starting 7 consumer threads
2018-11-09 15:06:38,271 [Thread-6] (com.pinterest.secor.main.ConsumerMain$1:65) ERROR Thread Thread[Thread-6,5,main] failed
java.lang.RuntimeException: Failed to initialize the consumer
	at com.pinterest.secor.consumer.Consumer.run(Consumer.java:90)
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: com.pinterest.secor.reader.LegacyKafkaMessageIterator
	at com.pinterest.secor.reader.KafkaMessageIteratorFactory.getIterator(KafkaMessageIteratorFactory.java:15)
	at com.pinterest.secor.reader.MessageReader.<init>(MessageReader.java:58)
	at com.pinterest.secor.consumer.Consumer.init(Consumer.java:69)
	at com.pinterest.secor.consumer.Consumer.run(Consumer.java:88)
Caused by: java.lang.ClassNotFoundException: com.pinterest.secor.reader.LegacyKafkaMessageIterator
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at com.pinterest.secor.reader.KafkaMessageIteratorFactory.getIterator(KafkaMessageIteratorFactory.java:11)
	... 3 more
2018-11-09 15:06:38,271 [Thread-9] (com.pinterest.secor.main.ConsumerMain$1:65) ERROR Thread Thread[Thread-9,5,main] failed
java.lang.RuntimeException: Failed to initialize the consumer
	at com.pinterest.secor.consumer.Consumer.run(Consumer.java:90)
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: com.pinterest.secor.reader.LegacyKafkaMessageIterator
	at com.pinterest.secor.reader.KafkaMessageIteratorFactory.getIterator(KafkaMessageIteratorFactory.java:15)
	at com.pinterest.secor.reader.MessageReader.<init>(MessageReader.java:58)
	at com.pinterest.secor.consumer.Consumer.init(Consumer.java:69)
	at com.pinterest.secor.consumer.Consumer.run(Consumer.java:88)
Caused by: java.lang.ClassNotFoundException: com.pinterest.secor.reader.LegacyKafkaMessageIterator
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at com.pinterest.secor.reader.KafkaMessageIteratorFactory.getIterator(KafkaMessageIteratorFactory.java:11)
	... 3 more
~~~~

From https://commons.apache.org/proper/commons-configuration/apidocs/org/apache/commons/configuration2/PropertiesConfiguration.html:
> If a key is used more than once, the values are appended like if they were on the same line separated with commas. Note: When the configuration file is written back to disk the associated PropertiesConfigurationLayout object (see below) will try to preserve as much of the original format as possible, i.e. properties with multiple values defined on a single line will also be written back on a single line, and multiple occurrences of a single key will be written on multiple lines. If the addProperty() method was called multiple times for adding multiple values to a property, these properties will per default be written on multiple lines in the output file, too. Some options of the PropertiesConfigurationLayout class have influence on that behavior.

Right now it seems that either I need to duplicate the default configuration completely, or apply magic[1].

--
[1] In my case "magic" will be to apply a local patch to the `secor.common.properties` file to set `kafka.message.iterator.className` and `kafka.client.className` before building.

--- a/pom.xml
+++ b/pom.xml
@@ -52,6 +52,9 @@
         <encoding>UTF-8</encoding>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <parquet.version>1.9.0</parquet.version>
+
+        <!-- Name of the directory below src/main/config that contains kafka-version-specific settings -->
+        <kafka.properties>kafka-legacy</kafka.properties>
     </properties>
 
     <distributionManagement>
@@ -356,6 +359,11 @@
         <resources>
             <resource>
                 <directory>src/main/config</directory>
+                <!-- Exclude the variable parts, these are copied by the respective profiles -->
+                <excludes>
+                    <exclude>kafka-legacy/**</exclude>
+                    <exclude>kafka-2.0.0/**</exclude>
+                </excludes>
             </resource>
         </resources>
         <testResources>
@@ -574,6 +582,31 @@
                     <target>1.8</target>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-resources-plugin</artifactId>
+                <version>3.1.0</version>
+                <executions>
+                    <execution>
+                        <id>copy-kafka-properties</id>
+                        <phase>process-resources</phase>
+                        <goals>
+                            <goal>copy-resources</goal>
+                        </goals>
+                        <configuration>
+                            <outputDirectory>${project.build.directory}/classes</outputDirectory>
+                            <resources>
+                                <resource>
+                                    <directory>src/main/config/${kafka.properties}</directory>
+                                    <includes>
+                                        <include>kafka.properties</include>
+                                    </includes>
+                                </resource>
+                            </resources>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
         </plugins>
     </build>
 
@@ -742,6 +775,9 @@
         </profile>
         <profile>
             <id>kafka-2.0.0</id>
+            <properties>
+                <kafka.properties>kafka-2.0.0</kafka.properties>
+            </properties>
             <dependencyManagement>
                 <dependencies>
                     <dependency>

--- a/src/main/config/kafka-2.0.0/kafka.properties
+++ b/src/main/config/kafka-2.0.0/kafka.properties
@@ -1,0 +1,24 @@
+#
+# Properties for using Kafka 2.0.0+
+#
+
+# Parameter which tells whether to extract Kafka message timestamp. This value is to be chose in case of 0.10.x kafka brokers.
+# Default value is false. Also specify `kafka.message.timestamp.className` as `com.pinterest.secor.timestamp.Kafka10MessageTimestamp`,
+# in case you are enabling this parameter as `true`.
+kafka.useTimestamp=false
+
+# Classname for the timestamp field you want to use. Default is `com.pinterest.secor.timestamp.Kafka10MessageTimestamp`
+# for 0.10 build profile. Basically, it will be `Kafka8MessageTimestamp` for 0.8 kafka and `Kafka10MessageTimestamp`
+# for 0.10 kafka. Fully classified names are `com.pinterest.secor.timestamp.Kafka8MessageTimestamp` and
+# `com.pinterest.secor.timestamp.Kafka10MessageTimestamp`.
+kafka.message.timestamp.className=com.pinterest.secor.timestamp.Kafka10MessageTimestamp
+
+# Classname for the message iterator you want to use. The message iterator determines what kind of consumer
+# secor will use to communicate with kafka. com.pinterest.secor.reader.LegacyKafkaMessageIterator uses
+# the old kafka consumer written scala. Its not recommended to use the legacy iterator with kafka version >= 1.0 since it
+# does not support the new broker protocols. You may face significant performance degradation on your brokers if you use it
+kafka.message.iterator.className=com.pinterest.secor.reader.SecorKafkaMessageIterator
+
+# Classname for the kafka client used by utility classes like PartitionFinalizer and Progress monitor.
+# Legacy client is kept for compatibility purposes and is deprecated
+kafka.client.className=com.pinterest.secor.common.SecorKafkaClient

--- a/src/main/config/kafka-legacy/kafka.properties
+++ b/src/main/config/kafka-legacy/kafka.properties
@@ -1,0 +1,24 @@
+#
+# Properties for using Kafka before 2.0.0
+#
+
+# Parameter which tells whether to extract Kafka message timestamp. This value is to be chose in case of 0.10.x kafka brokers.
+# Default value is false. Also specify `kafka.message.timestamp.className` as `com.pinterest.secor.timestamp.Kafka10MessageTimestamp`,
+# in case you are enabling this parameter as `true`.
+kafka.useTimestamp=false
+
+# Classname for the timestamp field you want to use. Default is `com.pinterest.secor.timestamp.Kafka10MessageTimestamp`
+# for 0.10 build profile. Basically, it will be `Kafka8MessageTimestamp` for 0.8 kafka and `Kafka10MessageTimestamp`
+# for 0.10 kafka. Fully classified names are `com.pinterest.secor.timestamp.Kafka8MessageTimestamp` and
+# `com.pinterest.secor.timestamp.Kafka10MessageTimestamp`.
+kafka.message.timestamp.className=com.pinterest.secor.timestamp.Kafka10MessageTimestamp
+
+# Classname for the message iterator you want to use. The message iterator determines what kind of consumer
+# secor will use to communicate with kafka. com.pinterest.secor.reader.LegacyKafkaMessageIterator uses
+# the old kafka consumer written scala. Its not recommended to use the legacy iterator with kafka version >= 1.0 since it
+# does not support the new broker protocols. You may face significant performance degradation on your brokers if you use it
+kafka.message.iterator.className=com.pinterest.secor.reader.LegacyKafkaMessageIterator
+
+# Classname for the kafka client used by utility classes like PartitionFinalizer and Progress monitor.
+# Legacy client is kept for compatibility purposes and is deprecated
+kafka.client.className=com.pinterest.secor.common.LegacyKafkaClient

--- a/src/main/config/secor.common.properties
+++ b/src/main/config/secor.common.properties
@@ -241,26 +241,7 @@ kafka.dual.commit.enabled=true
 # Possible values: "zookeeper" to read offset from zookeeper or "kafka" to read offset from kafka consumer topic
 kafka.offsets.storage=zookeeper
 
-# Parameter which tells whether to extract Kafka message timestamp. This value is to be chose in case of 0.10.x kafka brokers.
-# Default value is false. Also specify `kafka.message.timestamp.className` as `com.pinterest.secor.timestamp.Kafka10MessageTimestamp`,
-# in case you are enabling this parameter as `true`.
-kafka.useTimestamp=false
-
-# Classname for the timestamp field you want to use. Default is `com.pinterest.secor.timestamp.Kafka10MessageTimestamp`
-# for 0.10 build profile. Basically, it will be `Kafka8MessageTimestamp` for 0.8 kafka and `Kafka10MessageTimestamp`
-# for 0.10 kafka. Fully classified names are `com.pinterest.secor.timestamp.Kafka8MessageTimestamp` and
-# `com.pinterest.secor.timestamp.Kafka10MessageTimestamp`.
-kafka.message.timestamp.className=com.pinterest.secor.timestamp.Kafka10MessageTimestamp
-
-# Classname for the message iterator you want to use. The message iterator determines what kind of consumer
-# secor will use to communicate with kafka. com.pinterest.secor.reader.LegacyKafkaMessageIterator uses
-# the old kafka consumer written scala. Its not recommended to use the legacy iterator with kafka version >= 1.0 since it
-# does not support the new broker protocols. You may face significant performance degradation on your brokers if you use it
-kafka.message.iterator.className=com.pinterest.secor.reader.LegacyKafkaMessageIterator
-
-# Classname for the kafka client used by utility classes like PartitionFinalizer and Progress monitor.
-# Legacy client is kept for compatibility purposes and is deprecated
-kafka.client.className=com.pinterest.secor.common.LegacyKafkaClient
+include=kafka.properties
 
 # Secor generation is a version that should be incremented during non-backwards-compatible
 # Secor releases. Generation number is one of the components of generated log file names.

